### ISSUE TITLE
3D rendering seems to be completely broken on armv7

### ISSUE THREAD

**Issue by petm5** [minecraft-linux/mcpelauncher-manifest#852]
**Describe the bug**
On armv7, 3D rendering seems to be completely broken in recent versions of Minecraft (Starting around 1.19). The background of the menu screen is black, and inside worlds all you can see is a solid pinkish color that changes brightness when you try to look around. 3D views of your character in the Marketplace or in the inventory menu are just black.

All 2D sections of the UI like the hotbar and the settings menu seem to be rendering fine. 3D rendering works with no issues in versions prior to 1.19.

**To Reproduce**
1. Launch any version of the game after 1.18
2. Try to open any world
3. See error

**Expected behavior**
3D graphics should render correctly instead of showing a solid color.

**Desktop (please complete the following information):**
 - OS: Arch Linux armv7 via [junest](https://github.com/fsquillace/junest) (Alpine armv7 host)
 - Version: v0.10.0.716-AppImage-armhf (build 716)
 - Installation Type: AppImage

**Additional context**
This could be related to new versions using Render Dragon instead of the old rendering engine.

The version-specific bugs section in the README mentions that the armv7 port may not work correctly for newer Minecraft versions. Sorry for the noise if this is a known issue that is being worked on.

**Comment by ChristopherHX** [comment]
The arm32 port has low demand. All my 32bit only devices are too weak for 1.20.0+.
It took 8month+ for anyone filing an issue.

Yes it is possible to fix rendering, but then there are more things broken
- create world
- Select Touch input mode
- all other ore UI control won't load at all and the game freezes

The gfx problem is a calling convention mismatch hardfloat (linux) and softfloat (android).
If the compiler generates glue code then the gfx ingame works fine.

**Comment by petm5** [comment]
Interesting. UI menus and world generation seem to be working fine for me. Older versions of the game seem to be rendering perfectly fine with acceptable performance.

If there was an issue related to a mismatch between hardfloat and softfloat, I would expect newer versions of the game not to launch at all. Did the rendering engine switch to a different calling convention recently, or is the issue just related to the way the launcher handles these versions?

To potentially get rendering to work, am I looking at a compile with custom flags, or would parts of the launcher need an overhaul? Could the issue possibly be related to my kernel lacking support for a required feature?

**Comment by ChristopherHX** [comment]
The hardfloat and softfloat problem was solved for pre 1.18.30 while porting the launcher to armhf (the last time in 2020 using the opengl es wrappers of pre 2020).

> If there was an issue related to a mismatch between hardfloat and softfloat, I would expect newer versions of the game not to launch at all.

This would be the case for sysv_abi vs ms_abi, but not for armsf and armhf. The launcher contains wrappers for libc (compiler rt and libm is from ndk).

All functions, which don't have float/double parameter / return values are binary compatible between armhf and armsf.

armsf: floats are passed in int registers
armhf: floats are passed in float/fpu/neon registers

> Did the rendering engine switch to a different calling convention recently

No, the rendering engine swiched to previously unused non wrapped opengl es functions.

> would parts of the launcher need an overhaul?

Yes, the armhfsupport.cpp. I probably publish a patch for this soon. I got this working, but had issues you didn't see?

> Could the issue possibly be related to my kernel lacking support for a required feature?

No, it's 100 software side.


I doubt that you have an armv7 softfloat distro (maybe debian armel armv5 userland running an armv7 app), even if you have I have no idea if this launcher compiles.

**Comment by ChristopherHX** [comment]
Fixed in https://github.com/minecraft-linux/mcpelauncher-manifest/commit/1553d051fbd7d16677ad0d9616b9271930a23c5b.

BTW an appimage will be attached to the CI build as an artifact.

If you are lucky, then you can still create worlds even if I need something like 1.18.12 on armhf. Arm64 works for me.

**Comment by ChristopherHX** [comment]
> inside worlds all you can see is a solid pinkish color

You need a lot of RAM, a 1GB ram or weak GPU won't work with 1.20.0.
E.g. Raspberry Pi3 (armv7)
- black terrain, can be also missing textures / broken drivers
- missing item textures

I had to disable the kde sddm and start the game via `startx <pathtoAppImage> -p` to have enough video memory for the game.

With the change it works with 4GB ram in 32bit mode of a 64bit OS without graphical issues. This is on a raspberry Pi4.

> The background of the menu screen is black

This is fixed now, even on low ram devices, a raspberry pi3 sees and end of support October 2022 message due to it's small ram size.

**Comment by petm5** [comment]
I can confirm that 1.19 is working after the fix. I can create worlds fine on my Asus C201 with 2GB of RAM.

~Using the appimage from CI, version 1.20.0.01, which I already downloaded previously, does not launch. The launcher complains about the version not being supported. (It launched using the appimage from the releases tab with the same issues as 1.19.)~ Deleting the version and downloading it again fixed this.

I noticed that the launcher downloads an x86_64 apk if I try to run the latest version 1.20.1.02, which obviously does not work on my system. Not sure if this is expected or should be a new issue.
